File: projectforge-business/src/main/java/org/projectforge/reporting/Projekt.java
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2014 Kai Reinhard (k.reinhard@micromata.de)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.reporting;

import java.util.List;

import org.projectforge.business.fibu.ProjektStatus;


/**
 * Interface for reporting.
 * @author Kai Reinhard (k.reinhard@micromata.de)
 * @see org.projectforge.business.fibu.ProjektDO
 */
public interface Projekt
{
  /** Synthetischer, eindeutiger Datenbankschlüssel. */
  public Integer getId();

  /**
   * Ziffer 5-6 von KOST2 (00-99)
   */
  public int getNummer();

  public String getName();

  /**
   * @return Kunde or null if not exists.
   */
  public Kunde getKunde();

  /**
   * Nur bei internen Projekten ohne Kundennummer, stellt diese Nummer die Ziffern 2-4 aus 4.* dar.
   */
  public Integer getInternKost2_4();

  /**
   * Ziffer 2-4 von Kost2: Wenn Kunde gesetzt ist, wird die Kundennummer, ansonsten internKost2_4 zurückgegeben.
   */
  public Integer getBereich();

  /**
   * @see #getNummer
   */
  public Integer getTeilbereich();

  public ProjektStatus getStatus();

  public String getDescription();

  /** @return Kost2 1.-6. Stelle: 5.123.45 */
  public String getKost();

  public boolean isDeleted();

  /** @return Alle für das Projekt verfügbaren Kost2Arten (7.-8. Ziffer von Kost2). */
  public List<Kost2Art> getKost2Arts();
}

File: projectforge-rest/src/main/kotlin/org/projectforge/rest/dto/Projekt.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2020 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.rest.dto

import org.projectforge.business.fibu.ProjektDO
import org.projectforge.business.fibu.ProjektStatus

class Projekt(id: Int? = null,
              displayName: String? = null,
              var nummer: Int = 0,
              var name: String? = null,
              var identifier: String? = null,
              var status: ProjektStatus? = null,
              var kunde: Kunde? = null,
              var konto: Konto? = null,
              var task: Task? = null,
              var projektManagerGroup: Group? = null,
              var nummernkreis: Int? = null,
              var bereich: Int? = null,
              var kost2Arts: List<Kost2Art>? = null)
    : BaseDTODisplayObject<ProjektDO>(id, displayName = displayName) {

    /**
     * @see copyFromMinimal
     */
    constructor(src: ProjektDO): this() {
        copyFromMinimal(src)
    }

    override fun copyFrom(src: ProjektDO) {
        super.copyFrom(src)
        src.kunde?.let {
            this.kunde = Kunde(it)
        }
        src.konto?.let {
            this.konto = Konto(it)
        }
        src.task?.let {
            this.task = Task(it)
        }
    }

    val kost2ArtsAsString: String
        get() = kost2Arts?.joinToString { it.getFormattedId() } ?: ""

    fun transformKost2(allKost2Arts: List<org.projectforge.reporting.Kost2Art>?) {
        for (kost2 in allKost2Arts!!) {
            val kost2Art = Kost2Art()
            kost2Art.id = kost2.id
            kost2Art.name = kost2.name
            kost2Art.description = kost2.description
            kost2Art.fakturiert = kost2.isFakturiert
            kost2Art.projektStandard = kost2.isProjektStandard
            kost2Art.deleted = kost2.isDeleted
            kost2Art.selected = kost2.isSelected
            kost2Art.existsAlready = kost2.isExistsAlready
        }
    }
}
